File: 백준/Gold/16398. 행성 연결/행성 연결.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {

    static int[] parent;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        parent = new int[n];

        PriorityQueue<Edge> qu = new PriorityQueue<>();

        for (int i = 0; i < n; i++) {
            parent[i] = i;

            StringTokenizer st = new StringTokenizer(br.readLine());

            for (int j = 0; j < n; j++) {

                if (i == j) {
                    st.nextToken();
                    continue;
                }

                int cost = Integer.parseInt(st.nextToken());

                qu.offer(new Edge(i, j, cost));
            }
        }

        long ans = 0;

        while (!qu.isEmpty()) {

            Edge e = qu.poll();

            if (find(e.from) != find(e.to)) {
                union(e.from, e.to);
                ans += e.cost;
            }
        }

        System.out.println(ans);
    }

    private static void union(int a, int b) {
        a = find(a);
        b = find(b);
        if (a != b) {
            parent[b] = a;
        }
    }

    private static int find(int a) {
        if (parent[a] == a) {
            return a;
        }
        return parent[a] = find(parent[a]);
    }

    static class Edge implements Comparable<Edge> {

        int from, to, cost;

        public Edge(int from, int to, int cost) {
            this.from = from;
            this.to = to;
            this.cost = cost;
        }

        @Override
        public int compareTo(Edge o) {
            return this.cost - o.cost;
        }
    }
}

File: 백준/Gold/16398. 행성 연결/행성 연결.py
import sys
from queue import PriorityQueue

input = sys.stdin.readline

n = int(input())
qu = PriorityQueue()
parent = [0] * n

for i in range(n):
    parent[i] = i
    costs = list(map(int, input().split()))

    for j in range(i + 1, n):
        qu.put((costs[j], i, j))


def find(a):
    if parent[a] == a:
        return a
    parent[a] = find(parent[a])
    return parent[a]


def union(a, b):
    a = find(a)
    b = find(b)
    if a != b:
        parent[b] = a


ans = 0

while not qu.empty():
    cost, s, e = qu.get()

    if find(s) != find(e):
        union(s, e)
        ans += cost

print(ans)
